src/dev-utils/up.sh

```diff
@@ -9,45 +9,119 @@ CODE_UPSTREAM_REMOTE="upstream"
 CONFIG_DIR="$HOME/dev/Deployment"
 
 if [ $# != 2 ] ; then
-    echo Usage: $0 cluster_name version branch
+    echo Usage: $0 commit_hash cluster_name
     exit 1
 fi
 
-cluster_name=$1
-version=$2
-
-#(cd $CONFIG_DIR ; git fetch origin ; git merge origin/master)
-
-cd $CODE_BASE_DIR/$cluster_name
-
-git checkout . # make code path clean
-
-git fetch --all ; git checkout $CODE_UPSTREAM_REMOTE/$version -b $version
-
-cd src/ClusterBootstrap
-
-./ctl.py restorefromdir $CONFIG_DIR/$cluster_name
-
-# start to deploy
-
-for i in `echo restfulapi storagemanager repairmanager dashboard`
-do
-    ./ctl.py svc configupdate $i
-done
-
-./ctl.py kubectl label node `grep infra01: status.yaml  | sed 's/ *:*//g'` job-insighter=active
-
-for i in `echo RestfulAPI azure-blob-adapter dashboard email-sender gpu-reporter grafana init-container job-exporter job-insighter reaper RepairManager RepairManagerAgent RepairManagerEtcd StorageManager user-synchronizer watchdog`
-do
-    ./ctl.py docker push $i
-done
-
-for i in `echo restfulapi jobmanager monitor dashboard job-insighter logging repairmanager repairmanageragent storagemanager `
-do
-    ./ctl.py svc render $i
-    ./ctl.py svc stop $i
-    ./ctl.py svc start $i
-done
-
-watch -d -n 1 './ctl.py kubectl get pod --all-namespaces | grep -v Running'
-watch -d -n 1 './ctl.py kubectl get pod --all-namespaces | grep manager'
+version=$1
+cluster_name=$2
+
+function update_config() {
+    (
+        cd $CONFIG_DIR
+        git fetch origin
+        git merge origin/master
+    )
+}
+
+function update_code() {
+    (
+        cd $CODE_BASE_DIR/$cluster_name
+        git checkout . # make code path clean
+        git fetch --all ; git checkout $version
+    )
+}
+
+function restore_config() {
+    (
+        cd $CODE_BASE_DIR/$cluster_name/src/ClusterBootstrap
+        ./ctl.py restorefromdir $CONFIG_DIR/$cluster_name
+    )
+}
+
+
+function update_service_config() {
+    (
+        cd $CODE_BASE_DIR/$cluster_name/src/ClusterBootstrap
+        for i in $@ ; do
+            ./ctl.py svc configupdate $i
+        done
+    )
+}
+
+function push_docker_images() {
+    (
+        cd $CODE_BASE_DIR/$cluster_name/src/ClusterBootstrap
+        for i in $@ ; do
+            ./ctl.py docker push $i
+        done
+    )
+}
+
+function restart_services() {
+    (
+        cd $CODE_BASE_DIR/$cluster_name/src/ClusterBootstrap
+        for i in $@ ; do
+            ./ctl.py svc render $i
+            ./ctl.py svc stop $i
+            ./ctl.py svc start $i
+        done
+    )
+}
+
+function check() {
+    (
+        cd $CODE_BASE_DIR/$cluster_name/src/ClusterBootstrap
+        watch -d -n 1 './ctl.py kubectl get pod --all-namespaces | grep -v Running'
+        watch -d -n 1 './ctl.py kubectl get pod --all-namespaces | grep manager'
+    )
+}
+
+update_config
+
+update_code
+
+restore_config
+
+update_service_config `cat <(cat << EOF
+restfulapi
+storagemanager
+repairmanager
+dashboard
+EOF
+) | grep -v '#' `
```

Comment:
Why do we need `grep -v '#'`?

Reply:
I want to use this script in almost all upgrade. For example, if we only need to upgrade jobmanager, then we can comment out all unnessary steps and components.